src/solvers/dg.jl

```diff
@@ -756,15 +760,41 @@ function compute_coefficients!(u, func, t, mesh::AbstractMesh{1}, equations, dg:
     end
 end
 
-function compute_coefficients!(u, func, t, mesh::AbstractMesh{2}, equations, dg::DG,
+function compute_coefficients!(backend::Any, func, t, mesh::AbstractMesh{2}, equations,
+                               dg::DG,
                                cache)
+    @unpack node_coordinates = cache.elements
     @threaded for element in eachelement(dg, cache)
-        for j in eachnode(dg), i in eachnode(dg)
-            x_node = get_node_coords(cache.elements.node_coordinates, equations, dg, i,
-                                     j, element)
-            u_node = func(x_node, t, equations)
-            set_node_vars!(u, u_node, equations, dg, i, j, element)
-        end
+        compute_coefficients_element!(u, func, t, equations, dg, node_coordinates,
+                                      element)
+    end
+end
+
+function compute_coefficients!(backend::Backend, u, func, t, mesh::AbstractMesh{2},
+                               equations, dg::DG, cache)
+    nelements(dg, cache) == 0 && return nothing
+    # 1 cache not as argument
+    # 2 mesh not
+    @unpack node_coordinates = cache.elements
+    kernel! = compute_coefficients_kernel!(backend)
+    kernel!(u, func, t, equations, dg, node_coordinates,
+            ndrange = nelements(dg, cache))
+    return nothing
+end
+
+@kernel function compute_coefficients_kernel!(u, func, t, equations,
+                                              dg::DG, node_coordinates)
+    element = @index(Global)
+    compute_coefficients_element!(u, func, t, equations, dg, node_coordinates, element)
+end
```

Comment:
I would perhaps write this as a generalized `foreach_element` since I suspect that we will have this pattern quite often, but we can wait with generalizing until we have three cases of the same code.